File: alice.c
/*
 * Alice compress
 *
 * Attempt to pack 16-bit instructions using ALICE.exe compression.
 * Use a sliding window over instructions and bit shifting and OR to 
 * pack bits according to a dictionary.
 *
 * Copyright 2017 Donn Morrison donn.morrison@gmail.com
 *
 * Compile: gcc alice.c -o alice
 * 
 * TODO
 * Dictionary maker:
 *  - Where do the BitNumbers come from? --- Range registers
 *  - Why are there 7s prepended to some instructions? --- dunno, but correspond to bitNumber 13h
 *
 * Начало тела - body start
 * СжаТЬгй файл - combined file
 * таблица отображения - mapping table
 * словара - dictionary
 * до конца - to end
 *
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

//uint16_t buff[512];
int ptr = 0; // edi+10h
uint8_t edi12h = 0x07;
uint8_t edi1ch = 0;

void compress(uint8_t *buff, uint8_t bitNum, uint32_t fInstr)
{
//    printf("top of compress bitNum=0x%02x fInstr=0x%08x ptr=%d\n", bitNum, fInstr, ptr);
    uint8_t instrAndSeven0ch = (uint8_t)fInstr & 0xFF;

    if(bitNum != 0)
    {
        if(bitNum > 8)
        {
            bitNum -= 8;
            fInstr = fInstr >> 8;

            compress(buff, bitNum, fInstr);
            bitNum = 8;
        }

        uint8_t var1 = 1 << bitNum;
        uint8_t var2 = bitNum - edi12h;
        var1--;
        var1 = var1 & instrAndSeven0ch;
        var2--;
        instrAndSeven0ch = var2;

        if(var2 == 0)
        {
            // Rightmost tree
//            printf("0x%02x | 0x%02x = 0x%02x\n", buff[ptr], var1, buff[ptr] | var1);
            buff[ptr] = buff[ptr] | var1;
            ptr++;
            if(ptr == 0x100)
            {
                // call sub_190600 to reallocate
            }
            edi12h = 7;
            edi1ch = 1;
            return;

        }
        else if(var2 >> 7)
        {
            // Middle tree
            uint32_t var3 = var2;
            if(var2 >> 7)
                var3 = var2 | 0xFFFFFF00;
            var3 = -var3;
            var1 = var1 << var3;
//            printf("0x%02x | 0x%02x = 0x%02x\n", buff[ptr], var1, buff[ptr] | var1);
            buff[ptr] = buff[ptr] | var1;
            edi12h -= bitNum;
            edi1ch = 1;
            return;
        }
        else
        {
            // leftmost tree
            uint8_t var5 = var1 >> var2; // mov dl, bl; mov cl, al; shr dl, cl
//            printf("0x%02x | 0x%02x = 0x%02x\n", buff[ptr], var5, buff[ptr] | var5);
            buff[ptr] = buff[ptr] | var5; 
            ptr++;
            if(ptr == 0x100)
            {
                var2 = instrAndSeven0ch; // al
            }
            
            uint32_t var3 = var2; // esi
            if(var2 >> 7)
                var3 = var2 | 0xFFFFFF00;

            uint32_t var4 = 8 - var3;
            var1 = var1 << (uint8_t)(var4 & 0xFF);
            var4 = 0x07; // var4 = cl
//            printf("0x%02x | 0x%02x = 0x%02x\n", buff[ptr], var1, buff[ptr] | var1);
            buff[ptr] = buff[ptr] | var1;
            var4 = var4 - var2;
            edi12h = var4;
            edi1ch = 1;
            return;
        }
    }
}

void translate_bl_blx(uint16_t *buff, int count)
{
    // Careful, might be possible to overrun the buffer because
    // we increment i inside the loop at different stages
    int bl_count;
    int blx_count;
    for(int i=0; i<count; i++)
    {
        if((i+1) % 32 == 0) // Block size
        {
//            printf("skipping scan at %d\n", i+1);
            continue;
        }

        uint32_t instr, instr2, edi, esi;
        instr = buff[i]; // eax
//        printf("checking instruction 0x%02x\n", instr);
        edi = instr & 0xf800;

        if(edi == 0xf000)
        {
            instr2 = buff[i+1]; // edx
//            printf("found one! now checking next instruction 0x%02x\n", instr2);
            esi = instr2 & 0xf800;

            if(esi == 0xf800)
            {
                // BL instruction, translate
                instr2 &= 0x7ff;
                if(instr & 0x400)
                {
                    instr &= 0x7ff;
                    instr = instr << 0x0b;
                    instr += i;
                    instr2 += instr;
                    instr = instr2*2-0x7ffffe; // FIXME
                }
                else
                {
                    instr &= 0x7ff;
                    instr = instr << 0x0b;
                    instr += i;
                    instr2 += instr;
                    instr = instr2*2+2;
                }
                bl_count++;
                instr2 = instr >> 0x0c;
                instr = instr >> 1;
                instr &= 0x7ff;
                instr2 &= 0x7ff;

                instr |= esi;
                instr2 |= edi;

                printf("Translated 0x%04x to 0x%04x at %d\n", buff[i], instr2, i*2);
                printf("Translated 0x%04x to 0x%04x at %d\n", buff[i+1], instr, (i+1)*2);

                buff[i] = instr2;
                buff[i+1] = instr;

                i++; // increment i because we check for BLX below
            }
        }

        instr = buff[i]; // eax
//        printf("checking instruction 0x%02x\n", instr);
        edi = instr & 0xf800;

        if(edi == 0xf000)
        {
            instr2 = buff[i+1]; // edx
//            printf("found one! now checking next instruction 0x%02x\n", instr2);
            esi = instr2 & 0xf800;

            if(esi == 0xe800)
            {
                // BLX instruction, translate
                instr2 &= 0x7ff;
                if(instr & 0x400)
                {
                    instr &= 0x7ff;
                    instr = instr << 0x0b;
                    instr += i;
                    instr2 += instr;
                    instr = instr2*2-0x7ffffe; // FIXME
                }
                else
                {
                    instr &= 0x7ff;
                    instr = instr << 0x0b;
                    instr += i;
                    instr2 += instr;
                    instr = instr2*2+2;
                }
                blx_count++;
                instr2 = instr >> 0x0c;
                instr = instr >> 1;
                instr &= 0x7ff;
                instr2 &= 0x7ff;

                instr |= esi;
                instr2 |= edi;

                printf("Translated 0x%04x to 0x%04x at %d\n", buff[i], instr2, i*2);
                printf("Translated 0x%04x to 0x%04x at %d\n", buff[i+1], instr, (i+1)*2);

                buff[i] = instr2;
                buff[i+1] = instr;

                i++; // increment i because we check for BLX below
            }
        }
    }
}

int main(int argc, char *argv[])
{
    if(argc < 3)
    {
        printf("usage: alice <ALICE.bin> <magic.bin>\n");
        return 1;
    }

    struct stat st;
    stat(argv[1], &st);
    uint16_t *buff;
    uint8_t *encbuff;
    uint16_t *histcounts;
    uint8_t *magic;

    buff = malloc(st.st_size);
    if(!buff)
    {
        printf("malloc buff failed\n");
        return 1;
    }
    encbuff = malloc(st.st_size * 2);
    if(!encbuff)
    {
        printf("malloc encbuff failed\n");
        return 1;
    }
    histcounts = malloc(sizeof(uint16_t)*0xffff);
    if(!histcounts)
    {
        printf("malloc histcounts failed\n");
        return 1;
    }
    magic = malloc(0x10000*sizeof(uint8_t));
    if(!magic)
    {
        printf("malloc magic failed\n");
        return 1;
    }

    memset(buff, 0, st.st_size);
    memset(encbuff, 0, st.st_size*2);
    memset(histcounts, 0, sizeof(uint16_t)*0xffff);
    memset(magic, 0, sizeof(uint8_t)*0x10000);

    // 1. Read ALICE.bin bytes
    // 2. Translate BL/BLX
    // 3. Insert each instruction into binary search tree dictionary (self balancing?)
    // 4. Some kind of memory location copying loop
    // 5. Accumulator & shortest path
    // 6. Histogram
    // 7. Encode
    // 8. Write output

// Uncomment to read buff size bytes from file
    FILE *f = fopen(argv[1], "rb");
    int count=st.st_size/2;
    uint16_t instr;
    for(int j=0; j<st.st_size; j++)
    {
        fread(&buff[j], sizeof(uint16_t), 1, f);
    }

    printf("translating bl/blx instructions...\n");
    translate_bl_blx(buff, count);

    printf("buff:\n");
    for(int i=0; i<count; i++)
    {
        if(i%8 == 0 && i != 0)
            printf(" ");
        if(i%16 == 0 && i != 0)
            printf("\n");
        printf("0x%04x ", buff[i]);
    }
    printf("\n");

    FILE *fbuff = fopen("translated.bin", "wb");
    fwrite(buff, sizeof(uint16_t), count, fbuff);
    fclose(fbuff);

    printf("counting occurrences...\n");
    for(int i=0; i<count; i++)
    {
//        printf("%i found 0x%04x, %d -> %d\n", i, buff[i], histcounts[ buff[i] ],histcounts[ buff[i] ]+1);
        histcounts[ buff[i] ]++;
    }
    for(int i=0; i<0xFFFF; i++)
    {
        if(histcounts[ i ] != 0)
            printf("instr 0x%04x %d\n", i, histcounts[ i ]);
    }

    FILE *fmagic = fopen(argv[2], "rb");
    fread(magic, sizeof(uint8_t), 0x10000, fmagic);
    fclose(fmagic);

    printf("encoding instructions...\n");
    int start = 0x70000;
    for(int i=0; i<count; i++)
    {
        if(magic[ buff[i] ] == 0x13)
            compress(encbuff, 0x13, start + buff[i]);
        else
            compress(encbuff, magic[ buff[i] ], buff[i]);
    }

    printf("encbuff:\n");
    for(int i=0; i<st.st_size; i++)
    {
        if(i%8 == 0 && i != 0)
            printf(" ");
        if(i%16 == 0 && i != 0)
            printf("\n");
        printf("0x%02x ", encbuff[i]);
    }
    printf("\n");

    FILE *ebuff = fopen("translated_encoded.bin", "wb");
    fwrite(encbuff, sizeof(uint8_t), ptr, ebuff);
    fclose(ebuff);

    free(buff);
    free(encbuff);
    free(histcounts);
}

File: alice.py
import sys
import struct
from collections import Counter

'''
ALICE.exe encoder
    1. Read instructions from ALICE.bin
    2. Translate BL/BLX
    3. Generate dictionary
    4. Generate magic.bin bitnums, before_encoded.bin encoded instructions
    5. Compress - iterate through translated instructions (as index), encode
    6. Mapping table comes from bitpack() and blocksize, low byte unknown
    7. Postprocess

# magic.bin          - dictionary length array of lengths indexed by translated instruction
# before_encoded.bin - ranged instructions (32-bit form) in dictionary indexed by
#                      translated instruction
#
# TODO
# - DoMagic13And70000 sorting based on BST?
# - bitpack() end bytes (0x02) and padding
#
#
'''


def translate_bl_blx():
    global buff
    ptr = 0
    bl_count = 0
    blx_count = 0
    while ptr < len(buff)/2-1:
        if (ptr+1) % 32 == 0:
            ptr += 1
            continue

        instr = buff[ptr*2] | (buff[ptr*2+1] << 8)
        instr2 = buff[ptr*2+2] | (buff[ptr*2+3] << 8)

        if (instr & 0xf800) == 0xf000:
            if (instr2 & 0xf800) == 0xf800:
                upbits = 0xf800
                bl_count += 1
            elif (instr2 & 0xf800) == 0xe800:
                upbits = 0xe800
                blx_count += 1
            else:
                ptr += 1
                continue

            if instr & 0x400:
                v10 = 2 * (ptr + ((instr & 0x7ff) << 0x0b) + (instr2 & 0x7ff)) - 0x7ffffffe
            else:
                v10 = 2 * (ptr + ((instr & 0x7ff) << 0x0b) + (instr2 & 0x7ff)) + 0x00000002

            instr = (v10 >> 0x0c) & 0x7ff | 0xf000
            instr2 = (v10 >> 1) & 0x7ff | upbits

#            print("-translated 0x%04x to 0x%04x at 0x%x"%(buff[ptr*2] | buff[ptr*2+1] << 8, instr, ptr*2))
            buff[ptr*2] = instr & 0xff
            buff[ptr*2+1] = (instr >> 8) & 0xff
#            print("-translated 0x%04x to 0x%04x at 0x%x"%(buff[ptr*2+2] | buff[ptr*2+3] << 8, instr2, ptr*2+2))
            buff[ptr*2+2] = instr2 & 0xff
            buff[ptr*2+3] = (instr2 >> 8) & 0xff

            ptr += 1
        ptr += 1

    print("translated %d bl and %d blx instructions"%(bl_count*2, blx_count*2))

def bitpack(length, instr):
    print("in bitpack 0x%02x 0x%08x"%(length, instr))
    global buff, length_remain, ptr, count

    instr_part = instr & 0xff
    if length != 0:
        if length > 8:
            bitpack(length - 8, instr >> 8)
            length = 8

        length_part = length - length_remain - 1
        print("length_part: %d = %d - %d - 1"%(length_part, length, length_remain))

        if length_part < 0:
            # shift left
            print("shifting 0x%02x << %d = 0x%02x"%(instr_part, -length_part, (instr_part << -length_part) & 0xff))
            print("%d b %02x | %02x -> %02x"%(ptr, buff[ptr], (instr_part << -length_part) & 0xff, (buff[ptr] | (instr_part << -length_part) & 0xff) & 0xff))
            buff[ptr] |= (instr_part << -length_part) & 0xff
            print("length_remain: %d = %d - %d"%(length_remain - length, length_remain, length))
            length_remain -= length
        elif length_part == 0:
            # no shift
            print("no shift 0x%02x"%(instr_part))
            print("%d a %02x | %02x -> %02x"%(ptr, buff[ptr], instr_part, (buff[ptr] | instr_part) & 0xff))
            buff[ptr] |= instr_part
            ptr += 1
            buff.append(0x00)
            print("bitpack 1st buff.append")
            length_remain = 7
            print("length_remain: 7")
        elif length_part > 0:
            # shift right
            print("shifting 0x%02x >> %d = 0x%02x"%(instr_part, length_part, (instr_part >> length_part) & 0xff))
            print("%d c %02x | %02x -> %02x"%(ptr, buff[ptr], (instr_part >> length_part) & 0xff, (buff[ptr] | (instr_part >> length_part)) & 0xff))
            buff[ptr] |= (instr_part >> length_part) & 0xff
            ptr += 1
            buff.append(0x00)
            print("bitpack 2nd buff.append")
            # shift left into new byte
            print("shifting 0x%02x << %d = 0x%02x"%(instr_part, 8 - length_part, instr_part  << ((8 - length_part) & 0xff) & 0xff))
            print("%d d %02x | %02x -> %02x"%(ptr, buff[ptr], instr_part  << (8 - length_part & 0xff) & 0xff, (buff[ptr] | instr_part  << (8 - length_part & 0xff) & 0xff) & 0xff))
            buff[ptr] |= instr_part  << ((8 - length_part) & 0xff) & 0xff
            length_remain = 7 - length_part
            print("length_remain: %d = 7 - %d"%(length_remain, length_part))

# Read ALICE.bin

f = open(sys.argv[1], "rb")
buff = bytearray(f.read())
f.close()

# Translate BL and BLX instructions
ptr = 0
translate_bl_blx()

f = open("translated-py.bin", "wb")
f.write(buff)
f.close()

# Generate histogram
# Need to make sure it is sorted in the same way ALICE.exe sorts,
# that being that first by frequency, then for instructions of the
# same frequency, possibly by instruction value, first location 
# in ALICE.bin, or something else?
# Most likely ALICE.exe gets the order from the way the BST
# is traversed.

# Construct fake ALICE.bin with a desired histogram and try to match
# the output.
instrs=[bytes(buff[i*2:i*2+2]) for i in range(int(len(buff)/2))]
'''
hist=Counter(instrs)
shist=sorted(hist.items(), key=lambda x: (-x[1], x[0]))
shist_f=[freq for instr,freq in shist]
shist=[instr for instr,freq in shist]

# Generate magic vector, funked instructions dictionary (before_encode.bin)
range_regs = [0x0, 0x10, 0x50, 0xd0, 0x1d0, 0x3d0, 0xbd0, 0x1bd0]
#range_regs = [0x0, 0x02, 0x06, 0x0a, 0x0e, 0x12, 0x16, 0x1e]
codes = [0x07, 0x09, 0x0A, 0x0B, 0x0C, 0x0E, 0x0F, 0x13]
starts = [0x0, 0x40, 0x100, 0x300, 0x800, 0x2800, 0x6000, 0x70000]

magic = bytearray(0x10000)
fshist = dict()

for r in range(len(range_regs)-1):
    instrnr = 0
    for i in range(range_regs[r], range_regs[r+1]):
        instr_idx = struct.unpack("<H", shist[i])[0]
        print("0x%02x 0x%04x 0x%08x %d %d"%(codes[r], instr_idx, (instrnr | starts[r]), shist_f[i], i))
        # Magic
        magic[ instr_idx ] = codes[r]
        # Instr
        fshist[ instr_idx ] = instrnr | starts[r]
        instrnr += 1

offset = len(range_regs)-1
for i in range(range_regs[offset], len(hist)):
    instr_idx = struct.unpack("<H", shist[i])[0]
    print("0x%02x 0x%04x 0x%08x %d %d"%(codes[offset], instr_idx, (instr_idx | starts[offset]), shist_f[i], i))
    # Magic
    magic[ instr_idx ] = codes[offset]
    fshist[ instr_idx ] = instr_idx | starts[offset]

f=open("magic-py.bin", "wb")
f.write(magic)
f.close()

f=open("before_encode-py.bin", "wb")
for l,v in fshist.items():
    f.seek(l*4)
    f.write(struct.pack("<L", v))

f.close()
'''

# Try to encode!

if len(sys.argv) == 4:
    f=open(sys.argv[2], "rb")
    magic=bytearray(f.read())
    f.close()

    fshist=dict()
    f=open(sys.argv[3], "rb")
    for instr_idx in range(0x10000):
        instr = struct.unpack("<L", f.read(4))[0]
        fshist[ instr_idx ] = instr
    f.close()

ptr = 0
length_remain = 0x07
buff = bytearray([0x00])
count = 0
for instr in instrs:
    instr_idx = struct.unpack("<H", instr)[0]
    print("instr_idx %d 0x%04x total %d"%(instr_idx, instr_idx, len(fshist)))
    bitnum = magic[ instr_idx ]
    finstr = fshist [ instr_idx ]
    print("calling bitpack 0x%02x 0x%08x"%(bitnum, finstr))
    bitpack(bitnum, finstr)

    count += 1
    if count % 0x20 == 0:
        if count % 0x40 == 0:
            print("hit instruction blocksize 0x40 at ptr 0x%08x"%(ptr))
        if count > 0 and length_remain != 0x07:
            length_remain = 0x07
            ptr += 1
            buff.append(0x00)
            print("buff.append")
            # Do some stuff here...

# TODO Need to add some padding here

f=open("alice-py", "wb")
len = f.write(buff)
print("wrote alice-py %d bytes"%len)
f.close()
